Add `pivot_offset_ratio` property to Control

KoBeWi opened the pull request:
Spiritual salvage of #35493

This PR adds a `pivot_offset_ratio` property, which is `pivot_offset`, but expressed as 0-1 range mapped on Control's size. The new property is now serialized to scene instead of the old one. `pivot_offset` is now a wrapper for setting absolute offset (kind of like `rotation_degrees`) and will correctly get loaded if previously saved in scenes. Changing Control's size will preserve the offset ratio.

I added compat label, because you can no longer set pivot offset when control size is 0. Also changing control size will change the `pivot_offset`, which is technically new behavior.

KoBeWi committed:
Add pivot_offset_ratio property to Control

KoBeWi commented:
I added an alternate approach in a separate commit, which makes it more similar (same?) to #35493
The final pivot is now combination of the pixel offset and ratio offset.

For whatever reason changing ratio if the node is not rotated nor scaled won't update the editor viewport.

reduz commented:
To be entirely honest, I would make this a bit more similar to the other changes we made to control or even Nod3D:
* pivot_mode (exposed as enum), could have options (with beautiful icons we already have for other container stuff) like: 
   * PIVOT_TOP_LEFT
   * PIVOT_TOP_RIGHT
   * PIVOT_BOTTOM_LEFT
   * PIVOT_BOTTOM_RIGHT
   * PIVOT_CENTER
   * PIVOT_OFFSET
   * PIVOT_RATIO
* pivot_offset
* pivot_ratio

You can still save pivot_mode (that would most likely be ignored outside the editor) and pivot_offset like in this PR.

YuriSizov commented:
Since this is mostly useful for animations, I find that offset + offset ratio to be the most flexible and direct way to set it up. We can offer presets for the ratio as a QoL enhancement, but this can also be done later.

tokengamedev commented:
This is one of the desired feature. Animation of dialogs and other has to calculated otherwise to set the pivot position. This will really help.

Anutrix commented:
Any reason this is still stuck? Seems like a simple MR.
Been waiting for this for a while.

KoBeWi commented:
Rebased, squashed and fixed a bug. I also realized the user might want to use the editor's pivot tool to change the pivot as ratio, but that can be done in follow-up and probably requires additional discussion.

passivestar commented:
> Rebased, squashed and fixed a bug. I also realized the user might want to use the editor's pivot tool to change the pivot as ratio, but that can be done in follow-up and probably requires additional discussion.

Tested, works as expected, except for resizing

If there's non-default rotation or scale there's a problem (doesn't happen with the pixel-based pivot offset):

https://github.com/godotengine/godot/assets/60579014/9f4ccc36-8f1a-42c9-a953-fa0161e52b40

KoBeWi commented:
I checked it and I have no idea how to fix this :/
When you change size of the Control, the relative pivot will move and cause the whole Control to "move". The editor needs to account for the pivot position change somehow.

After further testing, it's partially a pre-existing bug. When you change pivot of a scaled Control, it will also behave in a weird way.

Meorge commented:
> When you change size of the Control, the relative pivot will move and cause the whole Control to "move". The editor needs to account for the pivot position change somehow.
> 
> After further testing, it's partially a pre-existing bug. When you change pivot of a scaled Control, it will also behave in a weird way.

Is this bug logged in a separate issue? To wrap my head around the problem, I tried to record it on 4.3.beta1:
https://github.com/godotengine/godot/assets/9957987/0cc3c785-58a8-40c4-8c37-b9eede392fa7

- **0:00** - Scale is `(1, 1)`, and I move the X pivot offset around. The PanelContainer itself doesn't move at all, but you can see the pivot point moving as expected.
- **0:12** - Scale is `(2, 1)`, and I move the X pivot offset around more. Now the PanelContainer *does* move. From what I can tell, it's trying to keep the pivot point at the true coordinates given to it (i.e., a pivot offset of `(50 px, 0 px)` is 50 px to the right of the parent's position, regardless of this object's scale). The PanelContainer is offset so that its visual *scaled* pivot offset matches the location of the true *unscaled* pivot offset.
  - This is made a bit clearer at **0:26** when I set the pivot offset to `(50 px, 0 px)`, `(100 px, 0 px)`, and `(0 px, 0 px)`, before resetting the scale to `(1, 1)`.

Does this sound like the behavior you were describing as the "pre-existing bug"? If so, what would the expected behavior be? I think the current behavior would also be what enables scaling around the pivot point. (I am rather tired at the moment, though, so I could very well be misunderstanding something 😅)

KoBeWi commented:
The issue is more prominent when you use the Pivot tool. The pivot position does not match cursor position.

https://github.com/godotengine/godot/assets/2223172/7601637f-dd62-40eb-b2f7-e1141deda84c

I think setting Pivot Offset should adjust the position to keep the Control in place.

Meorge commented:
To me this sounds more like an issue with the Pivot tool itself - it should factor in the object's scale when calculating where to place the pivot.

lfod1997 commented:
Do we stand a chance to have this feature in the near future? It solves so many problems and breaks almost nothing. 🥺

Mickeon reviewed (commented):
Docs are fine to me. This has been requested for quite a while, and the addition is straightforward.

Maybe the ratio itself doesn't need to be saved in the `data` struct? To save just a few bytes per Control? Mmm, maybe not.

Mickeon reviewed (approved):
Docs are fine to me. This has been requested for quite a while, and the addition is straightforward.

Maybe the ratio itself doesn't need to be saved in the `data` struct? To save just a few bytes per Control? Mmm, maybe not.

Edit: For some reason it sent the review 3 times in a row. I... don't know.

Mickeon reviewed (approved):
Docs are fine to me. This has been requested for quite a while, and the addition is straightforward.

Maybe the ratio itself doesn't need to be saved in the `data` struct? To save just a few bytes per Control? Mmm, maybe not.

KoBeWi commented:
> Maybe the ratio itself doesn't need to be saved in the data struct? To save just a few bytes per Control? Mmm, maybe not.

How it would be stored otherwise?

YeldhamDev reviewed (approved):
Tested, everything seems to work.
